clang/lib/CodeGen/CGExprScalar.cpp

```diff
@@ -1987,10 +1987,26 @@ Value *ScalarExprEmitter::VisitCastExpr(CastExpr *CE) {
     Value *Src = Visit(const_cast<Expr*>(E));
     llvm::Type *SrcTy = Src->getType();
     llvm::Type *DstTy = ConvertType(DestTy);
+    bool NeedAddrspaceCast = false;
     if (SrcTy->isPtrOrPtrVectorTy() && DstTy->isPtrOrPtrVectorTy() &&
         SrcTy->getPointerAddressSpace() != DstTy->getPointerAddressSpace()) {
-      llvm_unreachable("wrong cast for pointers in different address spaces"
-                       "(must be an address space cast)!");
+      // If we have the same address space in AST, which is then codegen'ed to
+      // different address spaces in IR, then an address space cast should be
+      // valid.
+      //
+      // This is the case for SYCL, where both types have Default address space
+      // in AST, but in IR one of them may be in opencl_private, and another in
+      // opencl_generic address space:
+      //
+      //   int arr[5]; // automatic variable, default AS in AST, private AS in IR
+      //   char* p = arr; // default AS in AST, generic AS in IR
+      //
+      if (E->getType().getAddressSpace() == DestTy.getAddressSpace()) {
+        NeedAddrspaceCast = true;
+      } else {
```

Comment:
Suggest rewriting to 
`if(E->getType().getAddressSpace() != DestTy.getAddressSpace()) 
    llvm_unreachable("wrong cast for pointers in different address spaces"
                         "(must be an address space cast)!");
NeedAddrspaceCast = true;`